pkg/updater/updater.go

```diff
@@ -432,3 +353,71 @@ func checkAvailableDiskSpace(fsDisk disk, paths ...string) error {
 	}
 	return nil
 }
+
+type requestKey int
+
+var requestStateKey requestKey
+
+// requestState represents the state of a task.
+type requestState struct {
+	Package string
+	ID      string
+	State   pbgo.TaskState
+	Err     *updaterErrors.UpdaterError
+}
+
+func newRequestContext(request remoteAPIRequest) context.Context {
+	return context.WithValue(context.Background(), requestStateKey, &requestState{
+		Package: request.Package,
+		ID:      request.ID,
+		State:   pbgo.TaskState_RUNNING,
+	})
+}
+
+func setRequestInvalid(ctx context.Context) {
+	state := ctx.Value(requestStateKey).(*requestState)
+	state.State = pbgo.TaskState_INVALID_STATE
+}
+
+func setRequestDone(ctx context.Context, err error) {
+	state := ctx.Value(requestStateKey).(*requestState)
+	state.State = pbgo.TaskState_DONE
+	if err != nil {
+		state.State = pbgo.TaskState_ERROR
+		state.Err = updaterErrors.From(err)
+	}
+}
+
+func (u *updaterImpl) refreshState(ctx context.Context) {
+	state, err := u.GetState()
+	if err != nil {
+		// TODO: we should report this error through RC in some way
+		log.Errorf("could not get updater state: %v", err)
+		return
+	}
+	requestState, ok := ctx.Value(requestStateKey).(*requestState)
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 Nitpick:** ‏Let's use `ok` to fail fast.

Reply:
Not sure if I see a way to fail fast here, we still have to get the state of all pacakges even if there is no request state 🤔

Reply:
I guess it boils down to if we want to freeze the state or reset it in case of context error. No strong opinion there, feel free to resolve if you prefer to reset